### PR TITLE
Added a GLTF loader.

### PR DESCRIPTION
**Pull request by VZout**
* Added a basic gltf loader.

### CODE REVIEW

File: src/model_loader_tinygltf.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+		std::vector<DirectX::XMFLOAT3> tanB(vtxCount, { 0, 0, 0 });
+
+		if (mesh_data->m_uvw.empty())
+		{
+			return { tanA, tanB };
+		}
+
+		// (1)
+		size_t indexCount = mesh_data->m_indices.size();
+		for (size_t i = 0; i < mesh_data->m_indices.size(); i += 3) {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: According to the style guide, a bracket should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+			tanA[i2] = tangent;
+
+			tanB[i0] = bitangent;
+			tanB[i1] = bitangent;
+			tanB[i2] = bitangent;
+
+		}
+
+		return { tanA, tanB };
+		// (2)
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Unused or debug code should be removed to keep the codebase clean.

File: src/model_loader_tinygltf.cpp [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+			// Get other attributes
+			for (const auto& attrib : primitive.attributes)
+			{
+				const auto attrib_accessor = tg_model.accessors[attrib.second];
+				const auto& buffer_view = tg_model.bufferViews[attrib_accessor.bufferView];
+				const auto& buffer = tg_model.buffers[buffer_view.buffer];
+				const auto data_address = buffer.data.data() + buffer_view.byteOffset + attrib_accessor.byteOffset;
+				const auto byte_stride = attrib_accessor.ByteStride(buffer_view);
+				const auto count = attrib_accessor.count;
+				if (attrib.first == "POSITION") {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Curly brackets should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+					mesh_data->m_positions.resize(position_buffer_offset + count);
+					mesh_data->m_colors.resize(position_buffer_offset + count);
+
+					memcpy(mesh_data->m_positions.data() + position_buffer_offset, data_address, count * byte_stride);
+					position_buffer_offset += count * byte_stride;
+				}
+				else if (attrib.first == "NORMAL") {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Curly brackets should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+					mesh_data->m_normals.resize(normal_buffer_offset + count);
+
+					memcpy(mesh_data->m_normals.data() + normal_buffer_offset, data_address, count * byte_stride);
+					normal_buffer_offset += count * byte_stride;
+				}
+				else if (attrib.first == "TEXCOORD_0") {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Curly brackets should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+		tinygltf::TinyGLTF loader;
+		std::string err;
+		std::string warn;
+		std::string path = std::string(model_path);
+
+		if (!loader.LoadASCIIFromFile(&tg_model, &err, &warn, path))
+		{
+			LOGC("TinyGLTF Parsing Failed");
+		}
+		//bool ret = loader.LoadBinaryFromFile(&model, &err, &warn, argv[1]); // for binary glTF(.glb)
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Unused code, might as well move it unless it's really convenient to keep it around.

File: src/model_loader_tinygltf.cpp [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+		if (!warn.empty()) {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Curly brackets should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+			LOGW("TinyGLTF Warning: {}", warn);
+		}
+
+		if (!err.empty()) {
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Curly brackets should be on a new line.

File: src/model_loader_tinygltf.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+		}
+
+		for (auto mesh : tg_model.meshes)
+		{
+			LoadMesh(model, tg_model, mesh);
+		}
+
+
+		// Load meshes
+		//LoadMeshes(model, tg_mode);
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Unused code, might as well remove it.

File: src/model_loader_tinygltf.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+		return model;
+	}
+
+	ModelData* TinyGLTFModelLoader::LoadModel(void* data, std::size_t length, std::string format)
+	{
+		auto model = new ModelData();
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Why go via "auto"? Might as well return immediately.

File: src/model_loader_tinygltf.hpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+		TinyGLTFModelLoader();
+
+		~TinyGLTFModelLoader() final;
+
+	protected:
+		ModelData* LoadModel(std::string_view model_path) final;
+		ModelData* LoadModel(void* data, std::size_t length, std::string format) final;
+
+	private:
+		//void LoadMeshes(ModelData* model, tinygltf::Model tg_model, tinygltf::Node node);
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.hpp]: Unused code, might as well remove it to keep the codebase clean.

File: src/model_loader_tinygltf.cpp (hunk 10 above)
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Also, is this ever deleted?

File: src/model_loader_tinygltf.cpp [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+		mat_data->m_albedo_texture_location = TextureLocation::NON_EXISTENT;
+		mat_data->m_normal_map_texture_location = TextureLocation::NON_EXISTENT;
+		mat_data->m_roughness_texture_location = TextureLocation::NON_EXISTENT;
+		mat_data->m_metallic_texture_location = TextureLocation::NON_EXISTENT;
+		mat_data->m_emissive_texture_location = TextureLocation::NON_EXISTENT;
+		mat_data->m_ambient_occlusion_texture_location = TextureLocation::NON_EXISTENT;
+
+		for (auto value : mat.values)
+		{
+			if (value.first == "baseColorTexture")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto img = tg_model.images[value.second.json_double_value.begin()->second];						
+
+				mat_data->m_albedo_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_albedo_texture = img.uri;
+			}
+			else if (value.first == "metallicRoughnessTexture")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto img = tg_model.images[value.second.json_double_value.begin()->second];
+
+				mat_data->m_metallic_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_metallic_texture = img.uri;
+				mat_data->m_roughness_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_roughness_texture = img.uri;
+			}
+			else if (value.first == "roughnessFactor")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto factor = value.second.Factor();
+				mat_data->m_base_roughness = factor;
+			}
+			else if (value.first == "metallicFactor")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto factor = value.second.Factor();
+				mat_data->m_base_metallic = factor;
+			}
+		}
+
+		for (auto value : mat.additionalValues)
+		{
+			if (value.first == "normalTexture")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto img = tg_model.images[value.second.json_double_value.begin()->second];
+
+				mat_data->m_normal_map_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_normal_map_texture = img.uri;
+			}
+			else if (value.first == "occlusionTexture")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto img = tg_model.images[value.second.json_double_value.begin()->second];
+
+				mat_data->m_ambient_occlusion_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_ambient_occlusion_texture = img.uri;
+			}
+			else if (value.first == "emissiveTexture")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+			{
+				auto img = tg_model.images[value.second.json_double_value.begin()->second];
+
+				mat_data->m_emissive_texture_location = TextureLocation::EXTERNAL;
+				mat_data->m_emissive_texture = img.uri;
+			}
+			else if (value.first == "emissiveFactor")
```
Comment (tntmeijs) [inline src/model_loader_tinygltf.cpp]: Can we turn this into constants?

File: src/model_loader_tinygltf.cpp (hunk 1 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: done

File: src/model_loader_tinygltf.cpp (hunk 2 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: done

File: src/model_loader_tinygltf.cpp (hunk 4 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: done

File: src/model_loader_tinygltf.cpp (hunk 5 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: done

File: src/model_loader_tinygltf.cpp (hunk 15 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: Me and clang format disagree.

File: src/model_loader_tinygltf.cpp (hunk 16 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: Me and clang format disagree.

File: src/model_loader_tinygltf.cpp (hunk 17 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: Me and clang format disagree.

File: src/model_loader_tinygltf.cpp (hunk 18 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: Me and clang format disagree.

File: src/model_loader_tinygltf.cpp (hunk 19 above)
Comment (VZout) [reply src/model_loader_tinygltf.cpp]: Me and clang format disagree.

File: resources/shaders/material_util.hlsl [hunk 20]
```diff
@@ -58,15 +58,16 @@ OutputMaterialData InterpretMaterialData(MaterialData data,
...
 
 	float4 albedo = lerp(float4(data.color, 1), material_albedo.Sample(s0, uv * data.albedo_uv_scale), use_albedo_texture);
-
-	float roughness = lerp(data.roughness, max(0.05f, material_roughness.Sample(s0, uv * data.roughness_uv_scale).x), use_roughness_texture);
-
+#ifdef COMPRESSED
+	float roughness = lerp(data.roughness, max(0.05f, material_roughness.Sample(s0, uv * data.metallic_uv_scale).y), use_roughness_texture);
+	float metallic = lerp(data.metallic, material_metallic.Sample(s0, uv * data.metallic_uv_scale).z, use_metallic_texture);
+#else
+	float roughness = lerp(data.roughness, max(0.05f, material_roughness.Sample(s0, uv * data.metallic_uv_scale).x), use_roughness_texture);
```
[the remark this answers is not in the document]
Comment (VZout) [reply resources/shaders/material_util.hlsl]: è cambiato

[the remark this answers is not in the document]
Comment (tntmeijs) [reply resources/shaders/material_util.hlsl]: ENGLISH PLEASE 😉

**Pull request merged**